Feat/prometheus-exporter : Add prometheus exporter and grafana dashboard for mailcow.

Hassanzadeh-sd opened the pull request:
<!-- _Please make sure to review and check all of these items, otherwise we might refuse your PR:_ -->

## Contribution Guidelines

* [x] I've read the [contribution guidelines](https://github.com/mailcow/mailcow-dockerized/blob/master/CONTRIBUTING.md) and wholeheartedly agree them

<!-- _NOTE: this tickbox is needed to fullfil on order to get your PR reviewed._ -->

## What does this PR include?

### Short Description

This PR adds a Prometheus exporter and Grafana dashboard for mailcow. Key changes:

- Prometheus Exporter: Collects mailcow metrics for monitoring. Requires enabling the mailcow API and generating an API key.
- /metrics Endpoint: Exposes metrics for Prometheus to scrape.
- Grafana Dashboard: Visualizes key mailcow metrics for easy monitoring.

How to use:

- Enable the Mailcow API and create a read-only API key.
- Configure the exporter with your mailcow host and API key.
- Import the provided Grafana dashboard to visualize the metrics.

This makes it easier to monitor your Mailcow instance with Prometheus and Grafana.

###  Affected Containers

nothing it is an override compose file.

## Did you run tests?

### What did you test?

I tested the Prometheus exporter and Grafana dashboard integration with my own Mailcow services. I verified that:

- The exporter successfully collected metrics from the mailcow API.
- The metrics were exposed correctly via the /metrics endpoint.
- Prometheus could scrape the metrics endpoint without errors.
- The Grafana dashboard visualized the collected metrics accurately and effectively.

The exporter and dashboard have been running in my environment and functioning as expected since testing.

### What were the final results? (Awaited, got)

All tests passed successfully. I expected the exporter to accurately collect and expose mailcow metrics in Prometheus format, and the Grafana dashboard to visualize these metrics effectively. The outcome matched my expectations, providing a functional monitoring solution for mailcow.

Hassanzadeh-sd committed:
add : readme for prometheus exporter configs

Hassanzadeh-sd committed:
add : grafana dashboard json file

Hassanzadeh-sd committed:
add: prometheus exporter service on docker-compose.override.yml

Hassanzadeh-sd commented:
@DerLinkman 

Hello again,

I have restructured my pull request as per your suggestions and in accordance with the contribution guidelines. I appreciate your feedback, and I would be grateful if you could take a moment to review the changes.

Thank you for your assistance!

DerLinkman commented:
Hi,

first of all, PR looks fine so far, thanks for that.

But i don't know how i should feel regarding the position of this PR.

Don't get me wrong i appreciate the work you did but the explainations on how to setup the stuff should better be put into our docs instead.

The prometheus exporter in an override.yml example file can be kept that is good. But the readme file on how to setup as well as the grafana dashboard template .json i don't feel it's not the right position.

What do you think?

DerLinkman commented:
Maybe you can add a seperate Docs section similar to the CheckMK stuff here:

https://docs.mailcow.email/third_party/checkmk/u_e-checkmk/

Hassanzadeh-sd committed:
migrate: doc files into docs.mailcow.email project

Hassanzadeh-sd commented:
thanks for your feedback, I removed my docs file in my pr branch and, I'll migrate them to docs.mailcow and send my new pr on that project asap.

Hassanzadeh-sd commented:
@DerLinkman 
I have seprated the documentation in to mailcow docs project and this is that PR:

https://github.com/mailcow/mailcow-dockerized-docs/pull/812

MAGICCC commented:
> Hi,
> 
> first of all, PR looks fine so far, thanks for that.
> 
> But i don't know how i should feel regarding the position of this PR.
> 
> Don't get me wrong i appreciate the work you did but the explainations on how to setup the stuff should better be put into our docs instead.
> 
> The prometheus exporter in an override.yml example file can be kept that is good. But the readme file on how to setup as well as the grafana dashboard template .json i don't feel it's not the right position.
> 
> What do you think?

Actually I second that it should be in the docs so users can install it when they want it. I gonna check it on the weekend

Hassanzadeh-sd commented:
hi, I would appreciate it if you could update me on whether you've had a chance to review my changes. Are there any updates or additional details required from me to move forward?

@DerLinkman

Hassanzadeh-sd commented:
please check this PR :)

@DerLinkman

DerLinkman commented:
So far the PR here looks fine. Just contacted the dev from mailcow-exporter if he would like to move his repo into the mailcow organization to be easier to find for users and to make it more official then using a third party image (if you know what i mean). I'll wait upon his reponse.

j6s commented:
IMO there needs to be a bit of documentation making clear, how to access the exporter in the internal network - also containing a warning not to expose the exporter publically since there is a bit of non-public data in there (e.g. list of all Mailboxes).

Since most people probably don't have their prometheus running in the same network as mailcow, adding some basic kind of authentication (e.g. basic-auth or predefined token) to the exporter may be a good idea

j6s commented:
I took this opportunity to re-engineer a couple of things in the exporter. It was originally built to be used on my NAS only, so security was not a big issue. Seeing, that we are now adding example of how to run it on the server itself means that most people will probably want to access the exporter over the internet.

 By default, a `token` is now required in order to access it from prometheus with an option to disable the token when accessing it internally. The image can now be found at `ghcr.io/mailcow/prometheus-exporter:2`. I do not provide a `latest` tag anymore, since I have come to believe that it is somewhat of an antipattern and makes setups more brittle for users and harder to support for developers.

milkmaker commented:
This pull request has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs.

Hassanzadeh-sd commented:
OK, let me do this, I'll fix it asap.

Commit:
add : security configs in prometheus exporter compose file

Hassanzadeh-sd commented:
I changed the image to the official Mailcow one with a specific version, and added the security feature you mentioned. @j6s 
https://github.com/mailcow/mailcow-dockerized/pull/6314/commits/40783cc1a381b81398476f05254b5028cccefc7f
in addtional I change the document side too.

j6s commented on helper-scripts/docker-compose.override.yml.d/PROMETHEUS_EXPORTER/docker-compose.override.yml:
```diff
@@ -0,0 +1,18 @@
+        MAILCOW_EXPORTER_HOST: "<your-mail-domain>" # Replace with your Mailcow hostname
+        MAILCOW_EXPORTER_API_KEY: "<your-API-Key>" # Replace with your API key
+        MAILCOW_EXPORTER_TOKEN: "<your-secure-token>" # Replace with your secure key
+        # MAILCOW_EXPORTER_TOKEN_DISABLE: "true" # uncomment only if safe and 
```
Seems the second half of the sentence is missing

Hassanzadeh-sd replied:
ok, I changed the comment sentence again.

j6s commented:
Apart from the comment, the change looks good to me (although I am lacking the ability to approve it)

Hassanzadeh-sd commented:
I have applied all the changes you suggested and also updated my Mailcow Docker Compose override configuration accordingly.

Commit:
add : explain more in my comment part in prometheus override compose file

Hassanzadeh-sd commented:
Kindly check the new changes,  @DerLinkman.

DerLinkman commented on mailcow-dockerized-docs:
Please remove

Hassanzadeh-sd replied:
I remove it. :)

Hassanzadeh-sd replied:
@DerLinkman

Hassanzadeh-sd committed:
remove : mailcow dockerized docs
